lib/GcInfo/GcInfo.cpp

```diff
@@ -0,0 +1,271 @@
+//===-------- include/gcinfo/gcinfo.cpp -------------------------*- C++ -*-===//
+//
+// LLILC
+//
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license.
+// See LICENSE file in the project root for full license information.
+//
+//===----------------------------------------------------------------------===//
+///
+/// \file
+/// \brief Implements the generation of CLR GCTables in LLILC
+///
+//===----------------------------------------------------------------------===//
+
+#include "earlyincludes.h"
+#include "GcInfo.h"
+#include "gcinfoencoder.h"
+#include "jitpch.h"
+#include "LLILCJit.h"
+#include "Target.h"
+#include "llvm/Object/StackMapParser.h"
+#include <map>
+#include <vector>
+#include <sstream>
+
+using namespace llvm;
+using namespace std;
+
+GCInfo::GCInfo(LLILCJitContext *JitContext, uint8_t *LLVMStackMapData,
+               uint8_t *CodeBlockStart) {
+  this->JitContext = JitContext;
+  this->LLVMStackMapData = LLVMStackMapData;
+  this->CodeBlockStart = CodeBlockStart;
+
+  this->Encoder = new GcInfoEncoder(JitContext->JitInfo, JitContext->MethodInfo,
+                                    new GcInfoAllocator());
+
+#if !defined(NDEBUG)
+  this->EmitLogs = JitContext->Options->LogGcInfo;
+#endif // !NDEBUG
+}
+
+void GCInfo::encodeHeader() {
+  if (EmitLogs) {
+    dbgs() << "GcTable for Function: " << JitContext->MethodName << "\n"
+           << "  Size: " << JitContext->HotCodeSize << "\n";
+  }
+
+  // TODO: Set Code Length accurately.
+  // https://github.com/dotnet/llilc/issues/679
+  // JitContext->HotCodeSize is the size of the allocated code block.
+  // It is not the actual length of the current function's code.
+  Encoder->SetCodeLength(JitContext->HotCodeSize);
+  Encoder->SetSizeOfStackOutgoingAndScratchArea(0);
+}
+
+void GCInfo::encodeLiveness() {
+  if (LLVMStackMapData == nullptr) {
+    return;
+  }
+
+  ArrayRef<uint8_t> StackMapContentsArray(
+      reinterpret_cast<const uint8_t *>(LLVMStackMapData),
+      JitContext->StackMapSize);
+
+#if defined(BIGENDIAN)
+  typedef StackMapV1Parser<support::big> StackMapParserType;
+#else
+  typedef StackMapV1Parser<support::little> StackMapParserType;
+#endif
+  StackMapParserType StackMapParser(StackMapContentsArray);
+
+  assert(StackMapParser.getNumFunctions() == 1 &&
+         "Expect only one function with GcInfo in the module");
+
```

Comment:
To clarify, this will need additional work on the LLVM side to associate records with functions in the StackMap section.